src/Microsoft.OData.Core/Metadata/EdmLibraryExtensions.cs

```diff
@@ -252,17 +254,129 @@ internal static IEnumerable<IEdmOperation> FilterBoundOperationsWithSameTypeHier
         }
 
         /// <summary>
-        /// Filters the operations by parameter names.
+        /// Filters the operations by parameter count.
         /// </summary>
-        /// <param name="functions">The operations.</param>
-        /// <param name="parameters">The list of non-binding parameter names to match.</param>
-        /// <param name="caseInsensitive">Whether case insensitive.</param>
+        /// <param name="operations">The operations.</param>
+        /// <param name="parameterCount">The count of non-binding parameter names to match.</param>
         /// <returns>The best matching operations based on parameters.</returns>
-        internal static IEnumerable<IEdmOperation> FindBestOverloadBasedOnParameters(this IEnumerable<IEdmOperation> functions, IEnumerable<string> parameters, bool caseInsensitive = false)
+        internal static IEnumerable<IEdmOperation> FindBestOverloadBasedOnParameterCount(this IEnumerable<IEdmOperation> operations, int parameterCount)
         {
             // The best match out of a list of candidates is the one that has the same number of (non-binding) parameters as specified.
-            IEnumerable<IEdmOperation> exactMatches = functions.Where(f => f.Parameters.Count() == parameters.Count() + (f.IsBound ? 1 : 0));
-            return exactMatches.Count() > 0 ? exactMatches : functions;
+            IEnumerable<IEdmOperation> exactMatches = operations.Where(f => f.Parameters.Count() == parameterCount + (f.IsBound ? 1 : 0));
+            return exactMatches.Count() > 0 ? exactMatches : operations;
+        }
+
+        /// <summary>
+        /// Filters the operations by parameter count.
+        /// </summary>
+        /// <param name="operations">The operations.</param>
+        /// <param name="parameterCount">The count of non-binding parameter names to match.</param>
+        /// <returns>The best matching operations based on parameters.</returns>
+        internal static IList<IEdmOperation> FindBestOverloadBasedOnParameterCount(this IList<IEdmOperation> operations, int parameterCount)
+        {
+            IList<IEdmOperation> exactMatches = null;
+
+            // The best match out of a list of candidates is the one that has the same number of (non-binding) parameters as specified.
+            for (int i = 0; i < operations.Count; i++)
+            {
+                if (operations[i].Parameters.Count() == parameterCount + (operations[i].IsBound ? 1 : 0))
+                {
+                    if (exactMatches == null)
+                    {
+                        exactMatches = new List<IEdmOperation>();
+                    }
+
+                    exactMatches.Add(operations[i]);
+                }
+            }
+
+            return exactMatches == null ? operations : exactMatches;
+        }

```

Comment:
Should return a single item and throw an exception if multiple operations match?

Reply:
Instead of throwing, I decided to change the name of the function to indicate filtering rather than finding the best overload. 

---
In reply to: [370893088](https://github.com/OData/odata.net/pull/1602#discussion_r370893088) [](ancestors = 370893088)